### PR TITLE
deduping nodes we’re getting from ZK

### PR DESCRIPTION
**Pull request by eishay**
this is very sensitive, I would like at least two of you to review it:
//cc @raymondng @ymatsuda @stkem 
related to the zk issue we discovered today

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceCluster.scala [hunk 1]
```diff
@@ -107,11 +107,40 @@ class ServiceCluster(val serviceType: ServiceType, airbrake: Provider[AirbrakeNo
       Some(leader)
   }
 
+  def deDuplicate(zk: ZooKeeperClient, instances: TrieMap[Node, ServiceInstance]): TrieMap[Node, ServiceInstance] = {
```
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceCluster.scala:110]: looks good. thanks @eishay!

I need to (obviously) brush up on this code -- question: do we need to update myInstance here when we dedup?

**Comment by stkem** [comment]
Looks good!
I'm a little concerned as to why this started to happen. I'll dig a little, or do you guys know already?

**Comment by eishay** [comment]
No idea

**Comment by raymondng** [comment]
Same here. Agreed that we should nail down the root cause.

**Comment by ymatsuda** [comment]
One possibility is a conflict between keepAlive and an invocation of onConnected handler.
We should remove keepAlive in the future. I will add a mutex for now.

**Comment by stkem** [comment]
Ah, that is also possible. That plus the reconnect issue we just discussed should cover it.

**Comment by ymatsuda** [comment]
I agree. Thanks!

**Pull request merged**